contracts/amm/pool/Joseph.sol

```diff
@@ -167,4 +187,133 @@ abstract contract Joseph is JosephInternal, IJoseph {
             return Constants.MAX_VALUE;
         }
     }
+
+    /// @dev Calculate redeem money
+    /// @param ipTokenAmount Amount of ipToken to redeem
+    /// @param exchangeRate Exchange rate of ipToken
+    /// @return redeemMoney Redeem money struct
+    function _calculateRedeemMoney(uint256 ipTokenAmount, uint256 exchangeRate)
+        internal
+        pure
+        returns (JosephTypes.RedeemMoney memory redeemMoney)
+    {
+        uint256 wadAssetAmount = IporMath.division(ipTokenAmount * exchangeRate, Constants.D18);
+
+        uint256 wadRedeemFee = IporMath.division(
+            wadAssetAmount * _getRedeemFeeRate(),
+            Constants.D18
+        );
+
+        uint256 redeemAmount = IporMath.convertWadToAssetDecimals(
+            wadAssetAmount - wadRedeemFee,
+            _getDecimals()
+        );
+
+        return
+            JosephTypes.RedeemMoney({
+                wadAssetAmount: wadAssetAmount,
+                wadRedeemFee: wadRedeemFee,
+                redeemAmount: redeemAmount,
+                wadRedeemAmount: IporMath.convertToWad(redeemAmount, _getDecimals())
+            });
+    }
+
+    function _rebalanceIfNeededBeforeRedeem(
+        IMiltonInternal milton,
+        uint256 wadMiltonErc20Balance,
+        uint256 vaultBalance,
+        uint256 wadOperationAmount
+    ) internal {
+        uint256 autoRebalanceThreshold = _getAutoRebalanceThreshold() * Constants.D21;
```

Comment:
Why do you multiply this by D21?
Maybe it could be done inside `_getAutoRebalanceThreshold()`

Reply:
ok

Reply:
I multiply by D21 because stored value is a multiplication of 1000, then in storage we can store value which represents higher value used  in code.

Reply:
I moved this multiplication to method.